[ADP-634] Run SMASH metadata fetching in batches of 15 concurrently

hasufell opened the pull request:
# Issue Number

ADP-634

# Overview

All requests already share the same [Manager](https://hackage.haskell.org/package/http-client-0.7.3/docs/Network-HTTP-Client.html#t:Manager). I'm not sure if that implicitly leads to proper HTTP pipelining when used with `forConcurrently`, but it seems to work: the sync time on testnet goes from 3 minutes down to ~30 seconds.

# Comments

## Open questions

1. do we also want to add a way for daedalus to check the sync-status? If so, how to define it? Syncing is continuous.
2. is this safe? We don't want to DoS SMASH
3. do we want to make the batch size a settings? We already have a database table for settings...

KtorZ committed:
prevent useless reward account queries when there's no accounts

  I noticed this while syncing an empty wallet for looking at stake pools. On a node which is also syncing, the wallet was spamming the reward account endpoint with an empty reward list. Interestingly, each request would take between 300ms and 500ms (which is quite a lot for an empty request...) and I believe, would also require quite some time / context-switch to the node. I think we ought to have a sort of debouncer here (though arguably, once synced, this only get triggered every 20s or so...); yet, a lot hanging fruit is simply to short-circuit everything when there's no account.

KtorZ committed:
rework pool metadata fetching to work using a bounded queue

  This works by using a bounded queue as a 'seat allocator'. Requests
  are only emitted in an asynchronous thread when there's a seat at
  the table.

  On my machine, the total time is cut from ~320s to ~50s. It also
  guarantees that there are never more than N requests in flight
  and does a best effort at keeping the number of requests in
  flight at any time close to N (this works because the time needed
  to fetch requests to send from the database is much smaller than
  the time needed to actually send and process requests.

  There's one seemingly important yet subtle change in this PR with
  regards to the fetching strategy.

  From:
  ```hs
  when (null refs || null successes) $ do
      traceWith tr $ MsgFetchTakeBreak blockFrequency
      threadDelay blockFrequency
  ```

  To:
  ```hs
  when (null refs) $ do
      traceWith tr $ MsgFetchTakeBreak blockFrequency
      threadDelay blockFrequency
  ```

  So, as of now, the wallet pauses for a whole 20s when all requests to
  fetching metadata failed. I think the rationale was originally to
  prevent the wallet from just eating up CPU by trying to fetch only
  invalid pools... What this create in practice is simply more
  unnecessary delays. There's about 8.5K pool registrations currently on
  mainnet, and after a full sync, there are still ~2254 fetch attempts
  still-to-proceed... so that's a lot of occasions to get a `null
  successes` on a particular loop.

  There's already a backoff which comes with the fetch attempts which
  prevent the wallet from doing needless work anyway. With this little
  change, we should massively improve the time needed for updating
  metadata upon new registrations, such as the problem described in:
  https://github.com/input-output-hk/daedalus/issues/2359

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -723,43 +738,79 @@ monitorMetadata gcStatus tr sp db@(DBLayer{..}) = do
+        :: TBQueue (PoolId, StakePoolMetadataUrl, StakePoolMetadataHash)
+        -> IO ()
+    enqueueMetadata queue = forever $ do
         refs <- atomically (unfetchedPoolMetadataRefs 100)
```
Perhaps set this number of refs to be smaller than queueSize - maybe the same as numInFlight.

KtorZ replied:
Whoops, I should've removed the magic number here and use `queueSize` instead. Thing is, I don't want the queue to be empty if there's something to fetch. So ideally, this should be > numInFlight. The bounded queue will block on writes anyway if the queue is full.

rvl commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -723,43 +738,79 @@ monitorMetadata gcStatus tr sp db@(DBLayer{..}) = do
+            h <- readTBQueue queue
+            q <- catMaybes <$> replicateM (numInFlight - 1) (tryReadTBQueue queue)
+            pure (h:q)
+        forConcurrently_ refs $ \(pid, url, hash) -> do
```
This will still have uneven bursts of requests, won't it?
But if the fetch time is down to one minute, then it's not really a problem.

KtorZ replied:
I see what you mean, we could actually make the consumer send queries one-by-one (asynchronously) and actually use the queue's size to limit the number of requests in flight :thinking: ...

KtorZ replied:
Done :+1:

KtorZ replied:
Whoops. Actually no, it's not that simple because now `unfetchedPoolMetadataRefs` gets called multiple times in a row before it's correctly updated.

KtorZ replied:
Okay, now done properly. After a full run, there's no attempts with a retry_count greater than 3, which sounds right.

rvl reviewed (approved):
:+1: Nice.
The time taken to fetch all metadata is highly dependent on your ping to the smash server.
So a 1 minute fetch in Europe could be longer in Australia.
How long did it take before to fetch all metadata on your connection?

hasufell commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -723,43 +734,58 @@ monitorMetadata gcStatus tr sp db@(DBLayer{..}) = do
+                fetchFromRemote trFetch strategies manager pid url hash >>= \case
+                    Nothing ->
+                        atomically $ putFetchAttempt (url, hash)
+                    Just meta -> do
```
redundant `do`?

KtorZ replied:
Who cares :sweat_smile: ? If the compiler judges it redundant, it'll remove it.

hasufell replied:
I think hlint will bail?

KtorZ replied:
Not if asked gently: https://github.com/input-output-hk/cardano-wallet/blob/master/.hlint.yaml#L51

hasufell commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -723,43 +734,58 @@ monitorMetadata gcStatus tr sp db@(DBLayer{..}) = do
+        inFlights <- STM.atomically $ newTBQueue maxInFlight
+        forever $ do
+            refs <- atomically (unfetchedPoolMetadataRefs $ fromIntegral maxInFlight)
+            forM_ refs $ \(pid, url, hash) -> withAvailableSeat inFlights $ do
```
This is a pretty neat pattern. Unrelatedly, I'm wondering if this is achievable with streamly or conduit.

KtorZ replied:
I actually thought of using a conduit here, to stream things directly from the database and then, process them at the right path. Yet, that's much more work on the db side to do, and persistent API is quite unsatisfactory here (we use a custom raw SQL query behind the scene, so most of the persistent API is just unusable). Plus, we would need a "refreshable" stream, because after requests are processed, the set of metadata refs to fetch do actually change.

hasufell commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -723,43 +734,58 @@ monitorMetadata gcStatus tr sp db@(DBLayer{..}) = do
+                threadDelay blockFrequency
+      where
+        maxInFlight :: Natural
+        maxInFlight = 20
```
Probably doesn't matter for merging the PR, but did someone check other values? I'm not questioning that 20 is a reasonable default, but maybe there's more room to optimize? At which point will performance degrade or SMASH give up on us?

KtorZ replied:
I've tried: 10, 20 and 50. There seem to be almost no difference and they all oscillate between 40s and 60s for a entire refresh :man_shrugging: .. I went for 10 in the end.

KtorZ committed:
do not re-fetch metadata already in flight

  I noticed that after running the new implementation, some metadata would have a retry_count way too high (internally, we use an exponential backoff, so it'd already take several days to reach ~6 or 7.. found some with 20+ retries). Turns out it was because the refs to fetch are only updated _after_ successful or unsuccessful fetching and since fetching is asynchronous, it could happen that a single metadata was simply enqueued over and over until a response finally came out of the server to change the state. Yet, all other queued request would also count towards the retry_count... and would simply eat up all the retries in one go.
